Implements by-casebook full-text search for legal documents

varun-magesh opened the pull request:
This PR is the first-step in enabling full-text search through casebooks. It implements the `/casebooks/<id>/search view`, which displays a list of legal documents associated with casebook `<id>` and displays a search bar that searches the full text of all the listed documents. To implement this view, it full-text indexes all of the cases used by H2O. This functionality has more potential uses that are not fully explored in this PR.

In order to complete full-text search through resources, we need to:
- [ ] index text resources
- [ ] index link resources
- [ ] (optionally) index annotations
- [ ] integrate search functionality into casebook view UI.

This PR is standalone and complete --- it can be merged now, or it can be merged together with other changes that complete the larger project at hand.

Closes #1568; should add a more detailed issue to address the problems directly above.

varun-magesh committed:
Wrote doctests for fts index, search, and view

lizadaly commented on web/main/models.py:
```diff
@@ -880,6 +883,53 @@ def _search(cls, category, query=None, page_size=10, page=1, filters={}, facet_f
+        ...     {}
+        ... )
+        """
+        if isinstance(casebook, int):
+            casebook = Casebook.objects.get(id=casebook)
+        if not isinstance(casebook, Casebook):
+            raise ValueError("param casebook is not an integer id or a Casebook object!")
```
`casebook` should only be one kind of thing here—I'd even recommend considering adding type hints that could enforce that in the future.

I recommend making the function parameter `casebook_id: int` and not supporting the case where the parameter could be a casebook object at all.

lizadaly commented on web/main/models.py:
```diff
@@ -880,6 +883,53 @@ def _search(cls, category, query=None, page_size=10, page=1, filters={}, facet_f
+        if not isinstance(casebook, Casebook):
+            raise ValueError("param casebook is not an integer id or a Casebook object!")
+         
+        legal_doc_ids = [node.resource.id for node in casebook.contents.filter(resource_type="LegalDocument").prefetch_resources()]
```
I don't think this needs to change, but I suspect if you only need the ids it would be faster/less memory to implement this query yourself directly using something like https://docs.djangoproject.com/en/3.2/ref/models/querysets/#values-list

lizadaly commented on web/main/create_search_index.sql:
```diff
@@ -20,6 +20,26 @@ CREATE MATERIALIZED VIEW internal_search_view AS
         main_legaldocument c
     GROUP BY c.id
+UNION ALL
+    -- seperate category for fts
```
```suggestion
    -- separate category for full-text search
```

varun-magesh commented:
Oh no --- I totally mucked this PR by not checking out of the black PR properly. Let me repair this mess :sweat_smile:

varun-magesh committed:
Added casebook/<id>/search page to expose FTS

The page currently only shows full text search
through legal documents. We should go through and
highlight fragments as well.

varun-magesh committed:
Added full-text index and search for LegalDocument

varun-magesh commented:
Okay, this should be fixed now! All of Liza's feeedback has been incorporated.

varun-magesh committed:
Remove LegalDocument.content from search index

This makes the materialized view much smaller and potentially
faster.

lizadaly reviewed (approved):
I reviewed this branch locally and it works great—the URL is available if you know it but not linked from the UI.
